src/services/helpers.ts

```diff
@@ -16,6 +16,36 @@ export async function fetchWithValidator<T, O, I>(
 ): Promise<T> {
   const response = await fetch(requestInfo, init);
 
+  const json = await response.json();
+
+  if (!response.ok) {
+    throw new Error(json.message ?? "Fetch failed");
+  }
+
+  const decoded = validator.decode(json);
+  return fold(
+    () => {
+      throw new ValidationError(reporter(decoded).join(" "));
+    },
+    (value: T) => value
+  )(decoded);
+}
+
+async function timeout<Response>(ms: number, promise: any): Promise<Response> {
+  return new Promise<Response>(async (resolve, reject) => {
+    setTimeout(() => {
+      reject(new Error("Request timed out. Please try again."));
+    }, ms);
+    resolve(await promise);
+  });
+}
```

Comment:
Given that this is a generic function and not confined to only `fetch`ing, we can make a couple of improvements:
1. Use generics for the type rather than using `Response`, the param of promise also doesn't need to be `any`
2. Using a `TimeoutError` will be useful as we propagate the error up the stack. We can perhaps handle it differently in future. e.g. when timeout error occurs, we show a different message
3. Given that this function may be used in different contexts, we can further make it customisable by leaving the message as an optional param. Also, I suggest we change it to be more actionable: "Request timed out. Please check your internet connection and try again later."